src/game/state/board/board.js

```diff
@@ -74,4 +77,124 @@ export default class Board {
+
+                if(position.humanReadable != otherIdEntities[uniqueId].position.humanReadable) {
+                    differences.push(new Difference({
+                        source,
+                        changeType: "updated",
+                        key: "positon",
+                        payload: {
+                            from: thisIdEntities[uniqueId].position,
+                            to: otherIdEntities[uniqueId].position,
+                        },
+                    }));
+                }
+
+                return differences;
+            },
+        });
+
+        const thisIdLocations = new Set(Object.values(thisIdEntities).map(entity => entity.position.humanReadable));
+        const otherIdLocations = new Set(Object.values(otherIdEntities).map(entity => entity.position.humanReadable));
+
+        const thisLevel = this[levelName == "unit" ? "_entities" : "_floor"];
+        const otherLevel = otherBoard[levelName == "unit" ? "_entities" : "_floor"];
+
+        const thisLocations = Object.keys(thisLevel).filter(location => !thisIdLocations.has(location));
+        const otherLocations = Object.keys(otherLevel).filter(location => !otherIdLocations.has(location));
+
+        differences = differences.concat(diffKeys({
+            fromKeys: thisLocations,
+            toKeys: otherLocations,
+            onAdd: location => {
+                return [new Difference({
+                    source: new BoardSource(new Position(location), levelName),
+                    changeType: "added",
+                    payload: thisLevel[location],
+                })];
+            },
+            onRemove: location => {
+                return [new Difference({
+                    source: new BoardSource(new Position(location), levelName),
+                    changeType: "removed",
+                })];
+            },
+            onElse: location => {
+                const source = new BoardSource(new Position(location), levelName);
+                return thisLevel[location].difference(otherLevel[location], source);
+            },
+        }));
+
+        return differences;
+    }
+
+    /**
+     * Get an object mapping unique id -> entity for all entities in the level
+     * @param {*} levelName the level of the board (unit or floor) to get the id map for
+     */
+    _getEntityIdMap(levelName) {
+        const thisLevel = this[levelName == "unit" ? "_entities" : "_floor"];
+        let idMap = {};
+
+        for(const entity of Object.values(thisLevel)) {
+            const id = entity.getUniqueId();
+            if(id !== undefined) {
+                if(idMap[id] !== undefined) {
+                    throw new Error(`Dupicate entity ID ${id} (entity1 = ${idMap[id]}, entity2 = ${entity})`);
+                }
+
+                idMap[id] = entity;
+            }
+        }
+
+        return idMap;
+    }
+}
+
+
+export class BoardSource {
```

Comment:
He's a slightly more complex source for the board.  Note that it uses id not position.
```
// To be able to track moving entities we give some entities a unique ID that doesn't change across moves.
// We start by diffing all entities with IDs and then we diff the rest of the entities with the assumption
```
My original intent was to implement this off of the engine's stable branch so I was limited to existing information for tracking.  Ideally this could be done by adding globally unique IDs to entities so we can track them if they move, change players, or get renamed.